File: mobile/src/library/playlists.ts
import { randomHex } from '../random'
import { openLibraryDb, withLibraryTransaction } from './db'
import type { LocalSong } from './songs'

/**
 * Playlists and favourites, on the device (#219).
 *
 * The last thing the server still owned about the library. Since #216 the songs
 * are local, so a favourite was a *server* row pointing at a song id the server
 * may never have heard of — and a device-fetched song could not be favourited
 * or added to a playlist at all.
 *
 * ## The invariants, and who enforces them now
 *
 * The backend enforced these; the schema does it here, which is stronger:
 *
 * - **Positions are dense and 0-based.** Removing renumbers.
 * - **Favourites is a playlist with `kind = 'favourites'`**, created by the
 *   migration, and it cannot be renamed or deleted.
 * - **A song appears at most once in a playlist**, which is a UNIQUE index
 *   rather than a check anybody has to remember.
 */

const ID_BYTES = 16
/**
 * The `kind` of the one playlist the library maintains itself.
 *
 * Exported since #324. There used to be a second copy of this string —
 * `FAVOURITES_KIND` in `src/api/playlists.ts` — documented as "the backend's
 * kind", and the two playlist screens imported *that* one to compare against
 * rows read from **this** device. It happened to work because the values
 * matched; it described a server the library stopped using at #159.
 */
export const FAVOURITES = 'favourites'

export interface LocalPlaylist {
  id: string
  name: string
  kind: string
  /** Set when a server playlist import produced this one (#225); null for a
   *  playlist made on the device. */
  server_playlist_id?: number | null
  created_at: string
  updated_at: string
  item_count: number
}

export interface LocalPlaylistItem {
  id: string
  position: number
  song: LocalSong
}

/** Refused because the backend refuses it too: favourites is not a user
 *  playlist and renaming or deleting it would leave the app with none. */
export class FavouritesIsProtected extends Error {
  constructor(action: string) {
    super(`Favourites cannot be ${action}`)
    this.name = 'FavouritesIsProtected'
  }
}

function now(): string {
  return new Date().toISOString()
}

export async function listPlaylists(): Promise<LocalPlaylist[]> {
  const db = await openLibraryDb()
  return db.getAllAsync<LocalPlaylist>(
    `SELECT p.*, (SELECT COUNT(*) FROM playlist_items i WHERE i.playlist_id = p.id) AS item_count
     FROM playlists p
     -- Favourites first, then newest. The server sorted by name; pinning
     -- favourites is what the screen actually wants and saves it re-sorting.
     ORDER BY (p.kind = 'favourites') DESC, p.created_at DESC`,
  )
}

export async function getPlaylist(playlistId: string): Promise<LocalPlaylist | null> {
  const db = await openLibraryDb()
  return db.getFirstAsync<LocalPlaylist>(
    `SELECT p.*, (SELECT COUNT(*) FROM playlist_items i WHERE i.playlist_id = p.id) AS item_count
     FROM playlists p WHERE p.id = ?`,
    [playlistId],
  )
}

/** The favourites playlist. Created by the migration, so this never has to
 *  decide whether it exists — the reason the backend seeds it too. */
export async function favouritesPlaylist(): Promise<LocalPlaylist> {
  const db = await openLibraryDb()
  const row = await db.getFirstAsync<LocalPlaylist>(
    `SELECT p.*, (SELECT COUNT(*) FROM playlist_items i WHERE i.playlist_id = p.id) AS item_count
     FROM playlists p WHERE p.kind = ?`,
    [FAVOURITES],
  )
  if (!row) throw new Error('The favourites playlist is missing')
  return row
}

export async function createPlaylist(name: string): Promise<string> {
  const db = await openLibraryDb()
  const id = randomHex(ID_BYTES)
  const timestamp = now()
  await db.runAsync(
    `INSERT INTO playlists (id, name, kind, created_at, updated_at) VALUES (?, ?, 'user', ?, ?)`,
    [id, name.trim(), timestamp, timestamp],
  )
  return id
}

/**
 * The local playlist for a playlist **import**, created once.
 *
 * Keyed so that reopening a finished import does not build a second copy — the
 * same job `songs.server_song_id` does, and the reason that column exists.
 *
 * ## Its own column, because an import id is not a playlist id (#308)
 *
 * This used to write the import's id into `playlists.server_playlist_id`, the
 * column the *handover* fills with a server **playlist** id. One UNIQUE column,
 * two unrelated numbering schemes: import #7 and server playlist #7 collided
 * onto one row. v6 adds `server_import_id` and this writes there.
 *
 * Rows written before v6 are left where they are — nothing records which kind
 * of id a value was, so they cannot be sorted out afterwards. An import that
 * already built a playlist may therefore build one more, once.
 */
export async function playlistForServerImport(importId: number, name: string): Promise<string> {
  const db = await openLibraryDb()
  const existing = await db.getFirstAsync<{ id: string }>(
    'SELECT id FROM playlists WHERE server_import_id = ?',
    [importId],
  )
  if (existing) return existing.id

  const id = randomHex(ID_BYTES)
  const timestamp = now()
  await db.runAsync(
    `INSERT INTO playlists (id, name, kind, created_at, updated_at, server_import_id)
     VALUES (?, ?, 'user', ?, ?, ?)`,
    [id, name.trim(), timestamp, timestamp, importId],
  )
  return id
}

/**
 * The local playlist for one **Google** playlist, created once (#106).
 *
 * The same job `playlistForServerImport` does, keyed on YouTube's own id — and
 * its own column, for the reason v6 exists: a Google playlist id is a string
 * YouTube issued, and the two `server_*` columns hold integers this project's
 * server issued. Sharing a UNIQUE column between two numbering schemes is the
 * fault v6 had to rebuild a table to undo.
 *
 * **Re-importing is the point.** A private playlist is a living list, so
 * running the import again to pick up what has been added since is ordinary
 * use — and it must fill in the playlist already on the phone rather than
 * leave two named the same. The name is *not* rewritten here: renaming the
 * local copy is the user's to do, and a re-import is not a request to undo it.
 */
export async function playlistForGooglePlaylist(
  googlePlaylistId: string,
  name: string,
): Promise<string> {
  const db = await openLibraryDb()
  const existing = await db.getFirstAsync<{ id: string }>(
    'SELECT id FROM playlists WHERE google_playlist_id = ?',
    [googlePlaylistId],
  )
  if (existing) return existing.id

  const id = randomHex(ID_BYTES)
  const timestamp = now()
  await db.runAsync(
    `INSERT INTO playlists (id, name, kind, created_at, updated_at, google_playlist_id)
     VALUES (?, ?, 'user', ?, ?, ?)`,
    [id, name.trim(), timestamp, timestamp, googlePlaylistId],
  )
  return id
}

/**
 * The local playlist for a Bilibili favourites folder, made if it is not here.
 *
 * `playlistForGooglePlaylist`'s twin (v8, #492), and everything in that one's
 * docblock applies: its own column rather than a shared one, because a
 * `media_id` is a number Bilibili issued and the `server_*` columns hold
 * integers this project's server issued — sharing a UNIQUE column between two
 * numbering schemes is the fault v6 had to rebuild a table to undo.
 *
 * **Re-importing is the point.** A favourites folder is a living list, so
 * running the import again to pick up what has been added since is ordinary
 * use, and it must fill in the playlist already on the phone rather than leave
 * two named the same. The name is *not* rewritten: renaming the local copy is
 * the user's to do, and a re-import is not a request to undo it.
 */
export async function playlistForBilibiliFav(mediaId: string, name: string): Promise<string> {
  const db = await openLibraryDb()
  const existing = await db.getFirstAsync<{ id: string }>(
    'SELECT id FROM playlists WHERE bilibili_fav_id = ?',
    [mediaId],
  )
  if (existing) return existing.id

  const id = randomHex(ID_BYTES)
  const timestamp = now()
  await db.runAsync(
    `INSERT INTO playlists (id, name, kind, created_at, updated_at, bilibili_fav_id)
     VALUES (?, ?, 'user', ?, ?, ?)`,
    [id, name.trim(), timestamp, timestamp, mediaId],
  )
  return id
}

/**
 * The playlist a multi-part Bilibili video imports into (#575, schema v9).
 *
 * Find-or-create on `bilibili_video_id`, for the reason v7, v8 and this all
 * share: importing four parts today and the other twenty-nine next week must
 * fill in the same playlist rather than build a second one beside it.
 *
 * Keyed on the `bvid` rather than the name, because an uploader can rename a
 * video and a name match would then quietly start a new playlist.
 */
export async function playlistForBilibiliVideo(bvid: string, name: string): Promise<string> {
  const db = await openLibraryDb()
  const existing = await db.getFirstAsync<{ id: string }>(
    'SELECT id FROM playlists WHERE bilibili_video_id = ?',
    [bvid],
  )
  if (existing) return existing.id

  const id = randomHex(ID_BYTES)
  const timestamp = now()
  await db.runAsync(
    `INSERT INTO playlists (id, name, kind, created_at, updated_at, bilibili_video_id)
     VALUES (?, ?, 'user', ?, ?, ?)`,
    [id, name, timestamp, timestamp, bvid],
  )
  return id
}

export async function renamePlaylist(playlistId: string, name: string): Promise<void> {
  const playlist = await getPlaylist(playlistId)
  if (playlist?.kind === FAVOURITES) throw new FavouritesIsProtected('renamed')

  const db = await openLibraryDb()
  await db.runAsync('UPDATE playlists SET name = ?, updated_at = ? WHERE id = ?', [
    name.trim(),
    now(),
    playlistId,
  ])
}

export async function deletePlaylist(playlistId: string): Promise<void> {
  const playlist = await getPlaylist(playlistId)
  if (playlist?.kind === FAVOURITES) throw new FavouritesIsProtected('deleted')

  const db = await openLibraryDb()
  // Explicit rather than a cascade: SQLite honours foreign keys only when
  // `PRAGMA foreign_keys` is on, which is off by default and per-connection.
  await withLibraryTransaction(db, async () => {
    await db.runAsync('DELETE FROM playlist_items WHERE playlist_id = ?', [playlistId])
    await db.runAsync('DELETE FROM playlists WHERE id = ?', [playlistId])
  })
}

/**
 * Delete several playlists at once (#570).
 *
 * ## What this does not do
 *
 * **It does not touch the songs.** A playlist is a list of references; deleting
 * it removes the references and leaves the music on the device. That is already
 * true of {@link deletePlaylist} and it is the property most worth stating,
 * because "delete" means *gone* everywhere else in this app — the library's own
 * delete takes the audio with it, and the two verbs sit two taps apart.
 *
 * ## Favourites is refused for the whole batch
 *
 * Not skipped. A batch that quietly dropped one of the things it was asked to
 * delete and reported success would be worse than an error: the user would have
 * to count the rows to find out. The screen keeps Favourites out of the list it
 * offers, so reaching this is a bug rather than a user's mistake, and it should
 * behave like one.
 */
export async function deletePlaylists(playlistIds: readonly string[]): Promise<number> {
  if (playlistIds.length === 0) return 0

  for (const playlistId of playlistIds) {
    const playlist = await getPlaylist(playlistId)
    if (playlist?.kind === FAVOURITES) throw new FavouritesIsProtected('deleted')
  }

  const db = await openLibraryDb()
  let removed = 0
  for (const chunk of chunked(playlistIds, SQLITE_VARIABLE_LIMIT)) {
    const places = chunk.map(() => '?').join(',')
    await withLibraryTransaction(db, async () => {
      // Explicit rather than a cascade, for the reason the singular form gives:
      // SQLite honours foreign keys only when `PRAGMA foreign_keys` is on,
      // which is off by default and per-connection.
      await db.runAsync(`DELETE FROM playlist_items WHERE playlist_id IN (${places})`, [...chunk])
      const result = await db.runAsync(`DELETE FROM playlists WHERE id IN (${places})`, [...chunk])
      removed += result.changes ?? 0
    })
  }
  return removed
}

/** A playlist's songs, in order. */
export async function listPlaylistItems(playlistId: string): Promise<LocalPlaylistItem[]> {
  const db = await openLibraryDb()
  const rows = await db.getAllAsync<LocalSong & { item_id: string; position: number }>(
    `SELECT s.*, i.id AS item_id, i.position
     FROM playlist_items i
     JOIN songs s ON s.id = i.song_id
     WHERE i.playlist_id = ?
     ORDER BY i.position`,
    [playlistId],
  )
  return rows.map(({ item_id, position, ...song }) => ({ id: item_id, position, song }))
}

/**
 * Append songs, skipping any already there.
 *
 * Silently skipping rather than failing: adding a playlist's worth of songs
 * where two overlap should add the rest, not refuse the lot. The UNIQUE index
 * is what makes "already there" a fact rather than a race.
 */
export async function addSongsToPlaylist(playlistId: string, songIds: string[]): Promise<number> {
  const db = await openLibraryDb()
  let added = 0

  await withLibraryTransaction(db, async () => {
    const last = await db.getFirstAsync<{ next: number }>(
      'SELECT COALESCE(MAX(position) + 1, 0) AS next FROM playlist_items WHERE playlist_id = ?',
      [playlistId],
    )
    let position = last?.next ?? 0

    for (const songId of songIds) {
      const existing = await db.getFirstAsync<{ id: string }>(
        'SELECT id FROM playlist_items WHERE playlist_id = ? AND song_id = ?',
        [playlistId, songId],
      )
      if (existing) continue

      await db.runAsync(
        'INSERT INTO playlist_items (id, playlist_id, song_id, position) VALUES (?, ?, ?, ?)',
        [randomHex(ID_BYTES), playlistId, songId, position],
      )
      position += 1
      added += 1
    }

    if (added > 0) {
      await db.runAsync('UPDATE playlists SET updated_at = ? WHERE id = ?', [now(), playlistId])
    }
  })

  return added
}

/** Renumber from 0 with no gaps — the invariant the UNIQUE index protects. */
async function renumber(
  db: Awaited<ReturnType<typeof openLibraryDb>>,
  playlistId: string,
): Promise<void> {
  const rows = await db.getAllAsync<{ id: string }>(
    'SELECT id FROM playlist_items WHERE playlist_id = ? ORDER BY position',
    [playlistId],
  )
  // Two passes, out of the way and back. A single pass collides with itself:
  // moving item 3 to position 1 hits the row already there, and the UNIQUE
  // index rejects it rather than letting a duplicate through.
  for (const [index, row] of rows.entries()) {
    await db.runAsync('UPDATE playlist_items SET position = ? WHERE id = ?', [-(index + 1), row.id])
  }
  for (const [index, row] of rows.entries()) {
    await db.runAsync('UPDATE playlist_items SET position = ? WHERE id = ?', [index, row.id])
  }
}

export async function removePlaylistItems(playlistId: string, itemIds: string[]): Promise<void> {
  if (itemIds.length === 0) return
  const db = await openLibraryDb()

  await withLibraryTransaction(db, async () => {
    for (const itemId of itemIds) {
      await db.runAsync('DELETE FROM playlist_items WHERE id = ? AND playlist_id = ?', [
        itemId,
        playlistId,
      ])
    }
    await renumber(db, playlistId)
    await db.runAsync('UPDATE playlists SET updated_at = ? WHERE id = ?', [now(), playlistId])
  })
}

/**
 * Reorder by the complete list of item ids, in their new order.
 *
 * The whole list, as the server's endpoint demanded: anything less makes
 * "where does everything else go" ambiguous, and a partial list is far more
 * likely to be a bug than an intention.
 */
export async function reorderPlaylist(playlistId: string, itemIds: string[]): Promise<void> {
  const db = await openLibraryDb()

  await withLibraryTransaction(db, async () => {
    const current = await db.getAllAsync<{ id: string }>(
      'SELECT id FROM playlist_items WHERE playlist_id = ?',
      [playlistId],
    )
    const currentIds = new Set(current.map((row) => row.id))
    const sameSet =
      currentIds.size === itemIds.length && itemIds.every((itemId) => currentIds.has(itemId))
    if (!sameSet) {
      throw new Error('Reorder must list exactly the playlist’s current items')
    }

    for (const [index, itemId] of itemIds.entries()) {
      await db.runAsync('UPDATE playlist_items SET position = ? WHERE id = ?', [
        -(index + 1),
        itemId,
      ])
    }
    for (const [index, itemId] of itemIds.entries()) {
      await db.runAsync('UPDATE playlist_items SET position = ? WHERE id = ?', [index, itemId])
    }
    await db.runAsync('UPDATE playlists SET updated_at = ? WHERE id = ?', [now(), playlistId])
  })
}

/** The song ids in favourites, for drawing the hearts. */
export async function favouriteSongIds(): Promise<string[]> {
  const db = await openLibraryDb()
  const favourites = await favouritesPlaylist()
  const rows = await db.getAllAsync<{ song_id: string }>(
    'SELECT song_id FROM playlist_items WHERE playlist_id = ?',
    [favourites.id],
  )
  return rows.map((row) => row.song_id)
}

/**
 * Which playlists already contain a song (#231).
 *
 * What lets the picker say "you already added this" rather than letting someone
 * add it twice and find out afterwards. Favourites is included — it is a
 * playlist row like any other — and the picker filters it out for its own
 * reasons, so the filtering stays where the decision is.
 */
export async function playlistIdsForSong(songId: string): Promise<string[]> {
  const db = await openLibraryDb()
  const rows = await db.getAllAsync<{ playlist_id: string }>(
    'SELECT DISTINCT playlist_id FROM playlist_items WHERE song_id = ?',
    [songId],
  )
  return rows.map((row) => row.playlist_id)
}

export async function setFavourite(songId: string, favourite: boolean): Promise<void> {
  const favourites = await favouritesPlaylist()
  if (favourite) {
    await addSongsToPlaylist(favourites.id, [songId])
    return
  }

  const db = await openLibraryDb()
  const item = await db.getFirstAsync<{ id: string }>(
    'SELECT id FROM playlist_items WHERE playlist_id = ? AND song_id = ?',
    [favourites.id, songId],
  )
  if (item) await removePlaylistItems(favourites.id, [item.id])
}

/** Every playlist entry for a song — what deleting a song has to clean up. */
export async function removeSongFromAllPlaylists(songId: string): Promise<void> {
  const db = await openLibraryDb()
  const affected = await db.getAllAsync<{ playlist_id: string }>(
    'SELECT DISTINCT playlist_id FROM playlist_items WHERE song_id = ?',
    [songId],
  )

  await withLibraryTransaction(db, async () => {
    await db.runAsync('DELETE FROM playlist_items WHERE song_id = ?', [songId])
    // Leaving a gap where the song was would break the dense-position
    // invariant for every playlist it happened to be in.
    for (const row of affected) {
      await renumber(db, row.playlist_id)
    }
  })
}

/**
 * The same, for many songs at once (#569).
 *
 * ## Why this exists rather than a loop over the singular form
 *
 * `renumber` is the expensive half, and the singular version runs it once per
 * **song**. Deleting forty songs out of one playlist renumbered that playlist
 * forty times — forty passes over the same rows to reach the same answer.
 *
 * Here the affected playlists are collected first and renumbered **once each**,
 * which is the whole of the win when a selection came out of one playlist. That
 * is the common case: it is how someone clears a playlist they no longer want.
 *
 * Chunked because SQLite's default host-parameter limit is 999, and "select
 * all" on a real library is past it. The limit is per statement, so chunking is
 * enough.
 */
export async function removeSongsFromAllPlaylists(songIds: readonly string[]): Promise<void> {
  // No early return for an empty list: `chunked` yields nothing, so the loop is
  // already a no-op and a guard here would be code no test could distinguish
  // from its absence. `removeLocalSongs` has one because *its* early return is
  // observable — it skips calling this at all.
  const db = await openLibraryDb()

  for (const chunk of chunked(songIds, SQLITE_VARIABLE_LIMIT)) {
    const places = chunk.map(() => '?').join(',')
    const affected = await db.getAllAsync<{ playlist_id: string }>(
      `SELECT DISTINCT playlist_id FROM playlist_items WHERE song_id IN (${places})`,
      [...chunk],
    )

    await withLibraryTransaction(db, async () => {
      await db.runAsync(`DELETE FROM playlist_items WHERE song_id IN (${places})`, [...chunk])
      // Once per playlist rather than once per song. Leaving a gap would break
      // the dense-position invariant, exactly as it would for one song.
      for (const row of affected) {
        await renumber(db, row.playlist_id)
      }
    })
  }
}

/** SQLite's default `SQLITE_MAX_VARIABLE_NUMBER` is 999; these queries bind one
 *  parameter per id, and the margin costs nothing. */
const SQLITE_VARIABLE_LIMIT = 900

function* chunked<T>(items: readonly T[], size: number): Generator<T[]> {
  for (let i = 0; i < items.length; i += size) yield items.slice(i, i + size)
}

File: backend/app/routers/playlists.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.orm import Session, selectinload

from app.db import get_db
from app.installs import current_owner, owned_by, require_install
from app.models import Playlist, PlaylistItem, PlaylistKind, Song
from app.schemas import (
    FavouriteSongIds,
    Page,
    PlaylistCreate,
    PlaylistDetail,
    PlaylistItemCreate,
    PlaylistRead,
    PlaylistReorder,
    PlaylistUpdate,
)

FAVOURITES_NAME = "Favourites"

router = APIRouter(prefix="/playlists", tags=["playlists"])


def _get_playlist_or_404(playlist_id: int, db: Session, owner_install_id: int | None) -> Playlist:
    """Someone else's playlist answers 404, not 403 — a 403 confirms it exists."""
    playlist = db.scalar(
        select(Playlist).where(Playlist.id == playlist_id, owned_by(Playlist, owner_install_id))
    )
    if playlist is None:
        raise HTTPException(status_code=404, detail="Playlist not found")
    return playlist


def _load_detail(playlist_id: int, db: Session, owner_install_id: int | None) -> Playlist:
    """Load a playlist with its items and their songs in one go (avoids N+1 queries)."""
    playlist = db.scalar(
        select(Playlist)
        .where(Playlist.id == playlist_id, owned_by(Playlist, owner_install_id))
        .options(selectinload(Playlist.items).selectinload(PlaylistItem.song))
    )
    if playlist is None:
        raise HTTPException(status_code=404, detail="Playlist not found")
    return playlist


def _get_favourites(db: Session, owner_install_id: int | None, *, create: bool) -> Playlist | None:
    """The caller's favourites playlist, created on first use.

    Created lazily rather than seeded by the migration so an untouched library
    doesn't sprout a playlist nobody asked for. There is one *per owner*, not
    one globally — two key holders each get their own hearts.
    """
    playlist = db.scalar(
        select(Playlist).where(
            Playlist.kind == PlaylistKind.FAVOURITES, owned_by(Playlist, owner_install_id)
        )
    )
    if playlist is not None or not create:
        return playlist

    playlist = Playlist(
        name=FAVOURITES_NAME, kind=PlaylistKind.FAVOURITES, owner_install_id=owner_install_id
    )
    db.add(playlist)
    db.commit()
    db.refresh(playlist)
    return playlist


def _reject_if_favourites(playlist: Playlist) -> None:
    """Favourites is managed by the heart button, not by playlist CRUD."""
    if playlist.kind == PlaylistKind.FAVOURITES:
        raise HTTPException(status_code=409, detail="The favourites playlist can't be changed")


def _renumber(playlist: Playlist, db: Session) -> None:
    """Rewrite positions as a dense 0-based sequence in current order."""
    items = db.scalars(
        select(PlaylistItem)
        .where(PlaylistItem.playlist_id == playlist.id)
        .order_by(PlaylistItem.position)
    ).all()
    for index, item in enumerate(items):
        item.position = index


@router.post("", response_model=PlaylistRead, status_code=201)
def create_playlist(
    payload: PlaylistCreate,
    db: Session = Depends(get_db),
    owner_install_id: int = Depends(require_install),
) -> PlaylistRead:
    playlist = Playlist(name=payload.name, owner_install_id=owner_install_id)
    db.add(playlist)
    db.commit()
    db.refresh(playlist)
    return PlaylistRead(
        id=playlist.id,
        name=playlist.name,
        kind=playlist.kind,
        item_count=0,
        created_at=playlist.created_at,
        updated_at=playlist.updated_at,
    )


@router.get("", response_model=Page[PlaylistRead])
def list_playlists(
    db: Session = Depends(get_db),
    owner_install_id: int | None = Depends(current_owner),
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
) -> Page[PlaylistRead]:
    total = (
        db.scalar(
            select(func.count()).select_from(Playlist).where(owned_by(Playlist, owner_install_id))
        )
        or 0
    )

    # Count items per playlist in one grouped subquery rather than loading every
    # item just to call len() on it.
    counts = (
        select(PlaylistItem.playlist_id, func.count().label("item_count"))
        .group_by(PlaylistItem.playlist_id)
        .subquery()
    )
    rows = db.execute(
        select(Playlist, func.coalesce(counts.c.item_count, 0))
        .outerjoin(counts, counts.c.playlist_id == Playlist.id)
        .where(owned_by(Playlist, owner_install_id))
        .order_by(Playlist.created_at.desc(), Playlist.id.desc())
        .limit(limit)
        .offset(offset)
    ).all()

    return Page[PlaylistRead](
        items=[
            PlaylistRead(
                id=playlist.id,
                name=playlist.name,
                kind=playlist.kind,
                item_count=item_count,
                created_at=playlist.created_at,
                updated_at=playlist.updated_at,
            )
            for playlist, item_count in rows
        ],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/favourites", response_model=PlaylistDetail)
def get_favourites(
    db: Session = Depends(get_db),
    owner_install_id: int = Depends(require_install),
) -> Playlist:
    """The favourites playlist, created on first read so the UI always has one."""
    playlist = _get_favourites(db, owner_install_id, create=True)
    assert playlist is not None  # create=True always returns one
    return _load_detail(playlist.id, db, owner_install_id)


@router.get("/favourites/song-ids", response_model=FavouriteSongIds)
def get_favourite_song_ids(
    db: Session = Depends(get_db),
    owner_install_id: int | None = Depends(current_owner),
) -> FavouriteSongIds:
    """Ids only — enough to draw a heart on every row in a list view, without
    shipping the whole favourites playlist to render it."""
    playlist = _get_favourites(db, owner_install_id, create=False)
    if playlist is None:
        return FavouriteSongIds(song_ids=[])
    rows = db.scalars(
        select(PlaylistItem.song_id)
        .where(PlaylistItem.playlist_id == playlist.id)
        .order_by(PlaylistItem.position)
    ).all()
    return FavouriteSongIds(song_ids=list(rows))


@router.post("/favourites/items", response_model=PlaylistDetail, status_code=201)
def add_favourite(
    payload: PlaylistItemCreate,
    db: Session = Depends(get_db),
    owner_install_id: int = Depends(require_install),
) -> Playlist:
    """Heart a song. Idempotent — hearting twice is not an error, and does not
    add it twice (unlike ordinary playlists, where duplicates are allowed)."""
    if (
        db.scalar(select(Song).where(Song.id == payload.song_id, owned_by(Song, owner_install_id)))
        is None
    ):
        raise HTTPException(status_code=404, detail="Song not found")

    playlist = _get_favourites(db, owner_install_id, create=True)
    assert playlist is not None
    existing = db.scalar(
        select(PlaylistItem).where(
            PlaylistItem.playlist_id == playlist.id, PlaylistItem.song_id == payload.song_id
        )
    )
    if existing is None:
        count = (
            db.scalar(
                select(func.count())
                .select_from(PlaylistItem)
                .where(PlaylistItem.playlist_id == playlist.id)
            )
            or 0
        )
        db.add(PlaylistItem(playlist_id=playlist.id, song_id=payload.song_id, position=count))
        db.commit()
    return _load_detail(playlist.id, db, owner_install_id)


@router.delete("/favourites/items/{song_id}", status_code=204)
def remove_favourite(
    song_id: int,
    db: Session = Depends(get_db),
    owner_install_id: int | None = Depends(current_owner),
) -> None:
    """Un-heart a song. Keyed by song id, not item id — the heart is on the
    song, and the caller has no reason to know the playlist item's id."""
    playlist = _get_favourites(db, owner_install_id, create=False)
    if playlist is None:
        return

    item = db.scalar(
        select(PlaylistItem).where(
            PlaylistItem.playlist_id == playlist.id, PlaylistItem.song_id == song_id
        )
    )
    if item is None:
        return

    db.delete(item)
    db.flush()
    _renumber(playlist, db)
    db.commit()


@router.get("/{playlist_id}", response_model=PlaylistDetail)
def get_playlist(
    playlist_id: int,
    db: Session = Depends(get_db),
    owner_install_id: int | None = Depends(current_owner),
) -> Playlist:
    return _load_detail(playlist_id, db, owner_install_id)


@router.patch("/{playlist_id}", response_model=PlaylistRead)
def update_playlist(
    playlist_id: int,
    payload: PlaylistUpdate,
    db: Session = Depends(get_db),
    owner_install_id: int | None = Depends(current_owner),
) -> PlaylistRead:
    playlist = _get_playlist_or_404(playlist_id, db, owner_install_id)
    _reject_if_favourites(playlist)
    playlist.name = payload.name
    db.commit()
    db.refresh(playlist)

    item_count = (
        db.scalar(
            select(func.count())
            .select_from(PlaylistItem)
            .where(PlaylistItem.playlist_id == playlist.id)
        )
        or 0
    )
    return PlaylistRead(
        id=playlist.id,
        name=playlist.name,
        kind=playlist.kind,
        item_count=item_count,
        created_at=playlist.created_at,
        updated_at=playlist.updated_at,
    )


@router.delete("/{playlist_id}", status_code=204)
def delete_playlist(
    playlist_id: int,
    db: Session = Depends(get_db),
    owner_install_id: int | None = Depends(current_owner),
) -> None:
    """Delete the playlist. Its items go too; the songs themselves are untouched."""
    playlist = _get_playlist_or_404(playlist_id, db, owner_install_id)
    _reject_if_favourites(playlist)
    db.delete(playlist)
    db.commit()


@router.post("/{playlist_id}/items", response_model=PlaylistDetail, status_code=201)
def add_playlist_item(
    playlist_id: int,
    payload: PlaylistItemCreate,
    db: Session = Depends(get_db),
    owner_install_id: int | None = Depends(current_owner),
) -> Playlist:
    """Append a song to the end of the playlist. The same song may appear twice."""
    playlist = _get_playlist_or_404(playlist_id, db, owner_install_id)
    # A song you do not own cannot be added, or a playlist would become a way to
    # name someone else's rows.
    if (
        db.scalar(select(Song).where(Song.id == payload.song_id, owned_by(Song, owner_install_id)))
        is None
    ):
        raise HTTPException(status_code=404, detail="Song not found")

    next_position = (
        db.scalar(
            select(func.count())
            .select_from(PlaylistItem)
            .where(PlaylistItem.playlist_id == playlist.id)
        )
        or 0
    )
    db.add(PlaylistItem(playlist_id=playlist.id, song_id=payload.song_id, position=next_position))
    db.commit()
    return _load_detail(playlist_id, db, owner_install_id)


@router.delete("/{playlist_id}/items/{item_id}", status_code=204)
def remove_playlist_item(
    playlist_id: int,
    item_id: int,
    db: Session = Depends(get_db),
    owner_install_id: int | None = Depends(current_owner),
) -> None:
    playlist = _get_playlist_or_404(playlist_id, db, owner_install_id)
    item = db.get(PlaylistItem, item_id)
    if item is None or item.playlist_id != playlist.id:
        raise HTTPException(status_code=404, detail="Playlist item not found")

    db.delete(item)
    db.flush()
    _renumber(playlist, db)
    db.commit()


@router.put("/{playlist_id}/items", response_model=PlaylistDetail)
def reorder_playlist_items(
    playlist_id: int,
    payload: PlaylistReorder,
    db: Session = Depends(get_db),
    owner_install_id: int | None = Depends(current_owner),
) -> Playlist:
    """Set a new order. `item_ids` must be exactly this playlist's item ids."""
    playlist = _get_playlist_or_404(playlist_id, db, owner_install_id)
    items = db.scalars(select(PlaylistItem).where(PlaylistItem.playlist_id == playlist.id)).all()

    by_id = {item.id: item for item in items}
    if sorted(payload.item_ids) != sorted(by_id):
        raise HTTPException(
            status_code=400,
            detail="item_ids must contain exactly the ids currently in this playlist",
        )

    for position, item_id in enumerate(payload.item_ids):
        by_id[item_id].position = position
    db.commit()
    return _load_detail(playlist_id, db, owner_install_id)
